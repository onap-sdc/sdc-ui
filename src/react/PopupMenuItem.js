import React from 'react';
import PropTypes from 'prop-types';

class PopupMenuItem extends React.Component {
    render() {
        const { itemId, value, onClick, selected, disabled } = this.props;
        const additionalClasses = selected
            ? 'selected'
            : disabled
                ? 'disabled'
                : '';
        return (
            <li
                className={`sdc-menu-item ${additionalClasses}`}
                onClick={event => {
                    event.stopPropagation();
                    if (onClick && !disabled) {
                        onClick(itemId);
                    }
                }}>
                {value}
            </li>
        );
    }
}

PopupMenuItem.propTypes = {
    itemId: PropTypes.any,
    value: PropTypes.any,
    selected: PropTypes.bool,
    onClick: PropTypes.func,
    disabled: PropTypes.bool
};

PopupMenuItem.defaultProps = {
    selected: false,
    disabled: false
};

export default PopupMenuItem;
