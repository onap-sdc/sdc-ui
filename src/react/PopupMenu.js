import React from 'react';
import PropTypes from 'prop-types';
import PopupMenuItem from './PopupMenuItem';

class PopupMenu extends React.Component {
    render() {
        const {
            children = [],
            onMenuItemClick,
            position = {},
            relative
        } = this.props;
        const style = relative ? { left: position.x, top: position.y } : {};

        return (
            <ul
                className={`sdc-menu-list ${relative ? 'relative' : ''}`}
                style={style}>
                {React.Children.toArray(children).map((child, i) =>
                    React.cloneElement(child, {
                        onClick: child.props.onClick || onMenuItemClick,
                        key: i
                    })
                )}
            </ul>
        );
    }
}

PopupMenu.propTypes = {
    relative: PropTypes.bool,
    position: PropTypes.shape({
        x: PropTypes.number,
        y: PropTypes.number
    }),
    onMenuItemClick: PropTypes.func
};

PopupMenu.defaultProps = {
    relative: false
};

export const PopupMenuSeparator = () => <li className="separator" />;

PopupMenu.Separator = PopupMenuSeparator;
PopupMenu.Item = PopupMenuItem;
export default PopupMenu;
