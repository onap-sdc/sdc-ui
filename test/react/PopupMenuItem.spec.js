import React from 'react';
import PopupMenuItem from '../../src/react/PopupMenuItem.js';
import PopupMenu from '../../src/react/PopupMenu.js';

import renderer from 'react-test-renderer';
import {mount} from 'enzyme';

describe('PopupMenuItem', () => {
	test('check selected', () => {
		const menuItem = renderer.create(<PopupMenuItem itemId='1' value='item 1' selected/>).toJSON();
		expect(menuItem).toMatchSnapshot();
	});

	test('check disabled', () => {
		const menuItem = renderer.create(<PopupMenuItem itemId='2' value='item 2' disabled/>).toJSON();
		expect(menuItem).toMatchSnapshot();
	});

	test('check parent onclick invoked by child', () => {
		const mockFunc = jest.fn();
		const menu = mount(
			<PopupMenu onMenuItemClick={mockFunc}>
				<PopupMenuItem itemId='1' value='item 1'/>
			</PopupMenu>
		);
		expect(menu.find('li')).toHaveLength(1);
		menu.find('li').simulate('click');
		expect(mockFunc).toHaveBeenCalled();
	});

	test('check custom onclick invoked by child', () => {
		const mockParentFunc = jest.fn();
		const mockChildFunc = jest.fn();
		const menu = mount(
			<PopupMenu onMenuItemClick={mockParentFunc}>
				<PopupMenuItem itemId='1' value='item 1' onClick={mockChildFunc}/>
			</PopupMenu>
		);
		expect(menu.find('li')).toHaveLength(1);
		menu.find('li').simulate('click');
		expect(mockChildFunc).toHaveBeenCalled();
		expect(mockParentFunc).not.toHaveBeenCalled();
	});

	test('check no click handler if item is disabled', () => {
		const mockFunc = jest.fn();
		const menu = mount(
			<PopupMenu onMenuItemClick={mockFunc}>
				<PopupMenuItem itemId='1' value='item 1' disabled/>
			</PopupMenu>
		);
		expect(menu.find('li')).toHaveLength(1);
		menu.find('li').simulate('click');
		expect(mockFunc).not.toHaveBeenCalled();
	});
});