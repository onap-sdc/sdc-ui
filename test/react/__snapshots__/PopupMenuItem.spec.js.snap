// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopupMenuItem check disabled 1`] = `
<li
  className="sdc-menu-item disabled"
  onClick={[Function]}
>
  item 2
</li>
`;

exports[`PopupMenuItem check selected 1`] = `
<li
  className="sdc-menu-item selected"
  onClick={[Function]}
>
  item 1
</li>
`;
